### PR TITLE
Fix uncheck checkbox and submit input type hidden value

### PR DESCRIPTION
**Pull request by SOHELAHMED7**
**Current status**

When a form have following inputs:

```html
<input type="hidden" name="coffee" value="0">
<input type="checkbox" name="coffee" value="1" id="coffee-id" checked>
```

and when checkbox is unchecked and then if the form is submitted, value `'0'` of `coffee` is not sent.

Ideally it should.

This PR fix that issue.

Failing test: https://github.com/Codeception/lib-innerbrowser/commit/d56484c53467ef5e2abb2b4d2bcd778f36679a8a

@SamMousa 

Have any questions, feel free to ask

### CODE REVIEW

File: src/Codeception/Lib/InnerBrowser.php [hunk 1]
```diff
@@ -1063,11 +1063,19 @@ protected function getFormFor(Crawler $node)
...
         $fields = $form->all();
         foreach ($fields as $field) {
-            if ($field instanceof FileFormField || $field->isDisabled() || !$field->hasValue()) {
-                continue;
+            if ($field instanceof FileFormField || $field->isDisabled()
+                ||
+                !$field->hasValue()
+            ) {
+                // if unchecked a checkboc and if there is hidden input with same name to submit unchecked value
+                $hiddenInput = $formNodeCrawler->filter('input[type=hidden][name="'.$field->getName().'"]');
```
Comment (SamMousa) [inline src/Codeception/Lib/InnerBrowser.php:1075]: Why is this needed? Since we are looping over all form fields I would expect us to find the hidden input eventually, no special action required..
Can you explain?

Comment (SOHELAHMED7) [reply src/Codeception/Lib/InnerBrowser.php:1075]: Input type with hidden and with same name as of `coffee` is not selected:

```
<input type="hidden" name="coffee" value="0123">
```

When I iterate over all fields 

```php
$fields = $form->all();
foreach ($fields as $field) {
```

and checkbox is unchecked then I get `NULL` in the value of coffee. Then that null is filtered out by `!$field->hasValue()` in line https://github.com/Codeception/lib-innerbrowser/blob/1.4.1/src/Codeception/Lib/InnerBrowser.php#L1069


So I searched (filtered) explicitly for input type hidden

Comment (SamMousa) [reply src/Codeception/Lib/InnerBrowser.php:1075]: Clear. Is that a bug in the dom crawler then? :wink:

Comment (SOHELAHMED7) [reply src/Codeception/Lib/InnerBrowser.php:1075]: Possibly not.


I don't have concrete knowledge regarding your question but there are are 4 actors involved in it

 - HTTP protocol
 - browser implementing that protocol
 - Server software such as Apache, nginx
 - Language like PHP who talks with Apache etc

How above software handle that might differ from software to software

From StackOverflow see answer comment with ~138 upvote - https://stackoverflow.com/a/1992745

So my conclusion is that there is not a bug in the dom crawler

Software tend to grab the last input only. That is why I used `$hiddenInput->last()->attr('value');`

Comment (SamMousa) [reply src/Codeception/Lib/InnerBrowser.php:1075]: I understand your solution, but what I'm unclear on is why it is needed.
The actors you mentioned are not relevant here. We don't use HTTP protocol, server software, a browser or FastCGI.

All we do is parse HTML using Symfony's DOM Crawler. This crawler has a function `->all()` which should return all fields.
If we iterate over these fields we expect to find both the checkbox AND the hidden input.
So if we then go ahead and store these values when the form field is valid we should automatically find the hidden input at some point...

To add to this, the DOM crawler itself has a `getPHPValues()` that already supposedly does the exact same thing. Could you try testing if this function works for our purposes?
That way we remove all manual parsing from this module and instead trust the upstream implementation.

Comment (SOHELAHMED7) [reply src/Codeception/Lib/InnerBrowser.php:1075]: I checked `Symfony\Component\DomCrawler\Form::getPHPValues()` . `coffee` is not present there in the returned array. Also I tried `Form::has()` and `Form::get`. In that I get `coffee` but that is of `input type="checkbox"` not of hidden

Comment (SamMousa) [reply src/Codeception/Lib/InnerBrowser.php:1075]: I can confirm that Symfony is struggling with this as well. Currently they can't handle it properly since they store the form field as a map from field name to object... 

Let me think some more on this.

Comment (SamMousa) [inline src/Codeception/Lib/InnerBrowser.php:1075]: This will also return inputs that are `disabled`, meaning we need to add a check for that.

This makes it troubling; it essentially means that all of the fall through logic that should be (and partially is) in DOM Crawler needs to be reimplemented here...

Comment (SOHELAHMED7) [reply src/Codeception/Lib/InnerBrowser.php:1075]: For `disabled` main inputs (here `$field` var), `continue;`  statement will be executed . Also see conditions in above `if` statement. Agree with you that `filter()` here will grab also disabled input but in real world app it is very less likely to have input that is hidden+disabled.

Comment (SamMousa) [reply src/Codeception/Lib/InnerBrowser.php:1075]: Sure; but that's a trade off that needs to be clear. Also isn't it simple to alter the call to `filter()` to skip disabled inputs?

Comment (SOHELAHMED7) [reply src/Codeception/Lib/InnerBrowser.php]: Yes there is way to filter out disabled input. Did that in https://github.com/Codeception/lib-innerbrowser/pull/36/commits/39f3e82f014e6e00b0ad7b602d117f96a2ffe391

**Comment by Naktibalda** [comment]
Wouldn't it be better to check if Checkbox field is unchecked and then skip it.  

Browsers process form fields in a linear fashion (that's the point of putting hidden field before checkbox).  
Your patch introduces non-linear behaviour which is nothing like browser's and it will introduce a lot new bugs.

Could you check if skipping unchecked fields work?

Before:
```php
foreach ($fields as $field) {
	if ($field instanceof FileFormField || $field->isDisabled() || !$field->hasValue()) {
	    continue;
	}
	// ...
}
```

After:
```php
foreach ($fields as $field) {
	if ($field instanceof FileFormField || $field->isDisabled() || !$field->hasValue()) {
	    continue;
	}
	
	if ($field instanceof CheckboxField && $field->isChecked() === false) {
	    continue;
	}
	// ...
}
```

**Comment by SamMousa** [comment]
The issue is that the symfony component does not add all fields to the form... It essentially deduplicates  based on name too early

**Comment by Naktibalda** [comment]
Ah, it is a WONTFIX issue: https://github.com/symfony/symfony/issues/11689

**Comment by SamMousa** [comment]
Yeah, so the question is do we want to work around it via the solution proposed by @SOHELAHMED7 ?

**Comment by Naktibalda** [comment]
It seems that a better solution would be to stop using Symfony\Component\DomCrawler\Form, extract all fields using Crawler and do all processing in InnerBrowser code, but it is a lot of work.

So a simple fix like this will have to do for now.

File: src/Codeception/Lib/InnerBrowser.php [hunk 2]
```diff
@@ -1063,11 +1063,19 @@ protected function getFormFor(Crawler $node)
...
-            if ($field instanceof FileFormField || $field->isDisabled() || !$field->hasValue()) {
-                continue;
+            if ($field instanceof FileFormField || $field->isDisabled()
+                ||
+                !$field->hasValue()
+            ) {
+                // if unchecked a checkboc and if there is hidden input with same name to submit unchecked value
+                $hiddenInput = $formNodeCrawler->filter('input[type=hidden][name="'.$field->getName().'"]:not([disabled])');
+                if (!count($hiddenInput)) {
+                    continue;
```
Comment (Naktibalda) [inline src/Codeception/Lib/InnerBrowser.php:1077]: This new code is in the wrong `if` block - it applies to file fields and disabled fields as well.

I think that it should be in its own block with `if(!$field->hasValue())` condition.  
I looked it up and `$field->hasValue()` returns false only for checkboxes and radio buttons [when they aren't selected](https://github.com/symfony/dom-crawler/blob/v5.2.4/Field/ChoiceFormField.php#L40-L53), so it is good for this purpose.

Also it would be good to get `$fieldValue` as a string right here to simplify the code below. The else case would set `$fieldValue = $field->getValue();` and the changes below would be simple assignments.

File: src/Codeception/Lib/InnerBrowser.php [hunk 3]
```diff
@@ -1063,11 +1063,19 @@ protected function getFormFor(Crawler $node)
...
+                }
             }
             $fieldName = $this->getSubmissionFormFieldName($field->getName());
             if (substr($field->getName(), -2) === '[]') {
```
Comment (Naktibalda) [inline src/Codeception/Lib/InnerBrowser.php:1081]: Hidden field can have `[]` in field name too.

**Review by Naktibalda — CHANGES_REQUESTED** [review]

**Comment by SOHELAHMED7** [comment]
@Naktibalda 

Do you mean to say like this https://github.com/Codeception/lib-innerbrowser/pull/36/commits/0d03718ed097873d4835c4a0699a4af762010a26  ?

File: src/Codeception/Lib/InnerBrowser.php [hunk 4]
```diff
@@ -1063,20 +1063,39 @@ protected function getFormFor(Crawler $node)
...
+                // if unchecked a checkboc and if there is hidden input with same name to submit unchecked value
+                $hiddenInput = $formNodeCrawler->filter('input[type=hidden][name="'.$field->getName().'"]:not([disabled])');
+                if (!count($hiddenInput)) {
+                    continue;
+                } else {
+                    // there might be multiple hidden input with same name, but we will only grab last one's value
+                    $hiddenFieldValue = $hiddenInput->last()->attr('value');
+                }
+            }
+
```
Comment (Naktibalda) [inline src/Codeception/Lib/InnerBrowser.php:1084]: Almost there, fields with `[]` don't benefit from your change yet.

I think that it would be the best to get fieldValue at this point.
```php
if (!$field->hasValue()) {
   ...
        $fieldValue = $hiddenInput->last()->attr('value');
    }
}  else {
   $fieldValue = $field->getValue();
```

and replace `$field->getValue();` with `$fieldValue` in all places below.

`$values[$fieldName][] = $field->getValue();` to `$values[$fieldName][] = $fieldValue;`

and your new block would become simply `$values[$fieldName] = $fieldValue;`

Comment (SOHELAHMED7) [reply src/Codeception/Lib/InnerBrowser.php:1086]: Fixed in fd66e5d2f18d6b2c3facc93023e56c3b88a1b47d
